spec/app/ics-029-fee-payment/README.md

```diff
@@ -0,0 +1,203 @@
+- One direction works, even when one chain does not support concept of fungible tokens
+- Opt-in for each chain implementing this. eg. ICS27 with fee support on chain A could connect to ICS27 without fee support on chain B.
+- Standardized interface for each chain implementing this extension
+- Allow each chain/application to customize fee-handling logic
+- Permissionless relaying
+
+## Technical Specification
+
+### General Design
+
+In order to avoid extra packets on the order of the number of fee packets, as well as provide an opt-in approach, we
+store all fee payment info only on the source chain. The source chain is the one location where the sender can provide tokens
+to incentivize the packet. The fee distribution may be implementation specific and thus does not need to be in the ibc spec
+(just high-level requirements are needed in this doc).
+
+We require that the [relayer address is exposed to application modules](https://github.com/cosmos/ibc/pull/579) for
+all packet-related messages, so the modules are able to incentivize the packet relayer. `OnAcknowledgement`, `OnTimeout`,
+and `OnTimeoutClose` messages will therefore have the relayer address and be capable of sending escrowed tokens to such address.
+However, we need a way to reliably get the address of the relayer that submitted `OnReceivePacket` on the destination chain to
+the source chain. In fact, we need a *source address* for this relayer to pay out to, not the *destination address* that signed
+the packet.
+
+Given this, the flow would be:
+
+1. User/module submits a send packet on the `source` chain, along with some tokens and fee information on how to distribute them
+2. Relayer submits `OnReceivePacket` on the `destination` chain. Along with this message, the *forward relayer* will submit a `payToOnSource` address where payment should be sent.
+3. Destination application includes this `payToOnSource` in the acknowledgement (there are multiple approaches to discuss below)
+4. Relayer submits `OnAcknowledgement` which provides the *return relayer* address on the source chain, along with the `payToOnSource` address
+5. Source application can distribute the tokens escrowed in (1) to both the *forward* and the *return* relayers.
+
+Alternate flow:
+
+1. User/module submits a send packet on the `source` chain, along with some tokens and fee information on how to distribute them
+2. Relayer submits `OnTimeout` which provides their address on the source chain
+3. Source application can distribute the tokens escrowed in (1) to this relayer, and potentially return extra tokens to the original packet sender.
+
+### Fee details
+
+For an example implementation in the Cosmos SDK, we consider 3 potential fee payments, which may be defined. Each one may be
+paid out in a different token. Imagine a connection between IrisNet and the Cosmos Hub. They may define:
+
+- ReceiveFee: 0.003 channel-7/ATOM vouchers (ATOMs already on IrisNet via ICS20)
+- AckFee: 0.001 IRIS
```

Comment:
As ack and timeout are mutually exclusive, and sender is normally interested more in ack, maybe sender can only allocate some tokens and pay complete amount to a relayer that brings ack, and only portion of it if timeout is brought.

Reply:
Sure, but the example ethan has works better to show how escrowing will work. As the minimum necessary escrow is 0.003 channel-7/ATOM and 0.002 IRIS

Reply:
> As ack and timeout are mutually exclusive, and sender is normally interested more in ack, maybe sender can only allocate some tokens and pay complete amount to a relayer that brings ack, and only portion of it if timeout is brought.

For ICS20, I am most interested in Receive, and not really interested in Ack... unless if is a failure (to release my tokens). So far, most "failure cases" are really timeouts. So for ICS20, I would incentivize receive and timeout. The *forward relayer* would only relay the ack to claim the receive fee. (not my example here, but if we try to model some "rational actor")

Reply:
As a rational actor, I would set a high receive fee and just see if that's enough to get my receive and ack relayed by the same relayer. Maybe I would do a small ack fee.

I'd only set a timeout fee at all once the packet has already timed out

Reply:
> I'd only set a timeout fee at all once the packet has already timed out

You mean you can change a fee after sending? That would be a huge complexity increase (also on the relayer, which must re-check every packet even after seeing "too low fee"). There is no changes to increasing tx fees, and I would like to remove such an approach.

Reply:
Could you explain your reasoning here? I don't understand.

> You mean you can change a fee after sending? That would be a huge complexity increase

I only would allow the ability to increase a fee after sending. I don't think it's a huge complexity increase at all. All it would mean is the ability to provide a message that escrows more fee funds against a particular packetID.
In fact, my original proposal had the fee be a completely separate message that could be sent synchronously or asynchronously with the packet.
I do think most apps (and maybe in ICS-4) would like to set fee on Send, but I don't see a reason why those fees can't be increased if the packet is just not getting relayed by anyone.

From the blockchain standpoint, this would simply be implementing a message on the fee-module that allows you to set fees asynchronously from the packet send. The logic of the handler would simply call `SetFee`, which itself will add fees onto any existing escrow or set the escrow if it doesn't already exist.
Applications (or ICS-4) may set the initial fee synchronously during execution by calling the `SetFee` callback, but it may be asynchronously set through the ibc-fee handler itself.

From the relayer point of view, re-checking fees is not required. If you ignore something due to initially low fee, you lose out to smarter relayers. I don't think it's that difficult to create a smart relayer:
Emit events whenever the escrow amount for a packet is increased, relayers can monitor these to see if a packet is worth relaying (regardless of if they've checked before) and relay if they believe it is.

I don't think it's that complex from the blockchain or relayer view.

Additionally, I think it is a big UX boost for users. We want to make sure that users can get their packets relayed.

In the regular transaction case, if your tx gets rejected from mempool you can easily just resubmit with no issues. In the IBC case, if your packet is successfully committed but no relayers want to relay your packet, there's no recourse except to relay your own packet. Until then any funds (ICS-20 case) are stuck. From a user perspective, it means I have to make sure I have enough fees (probably overestimate) to ensure a good UX experience.

I think in the case where a user puts in zero fees or too low fees during packet send, I don't want their funds indefinitely stuck. Nor do I want to force them to set up a relayer and relay their own fees to recover their funds, just because they were hoping for an altruistic relay. Ideally, they wait for some time; realize no one's picking up their packet with those fees, and decide to bump fees in the hope that it gets relayed then.

This was discussed as a strength of the proposal here: https://github.com/cosmos/ibc/issues/578#issuecomment-851622881

Reply:
> I only would allow the ability to increase a fee after sending. I don't think it's a huge complexity increase at all. All it would mean is the ability to provide a message that escrows more fee funds against a particular packetID.
In fact, my original proposal had the fee be a completely separate message that could be sent synchronously or asynchronously with the packet.

This is not much complexity increase on the blockchain side. And only allowing increases makes it safe for relayers.

The complexity increase is that relayers must be conscious of all future changes to this packet state. No longer checking the packet the first time and putting in "relay" or "ignore" buckets, but they must re-evaluate all "ignores" in future iterations.

This could get complex.

> From the relayer point of view, re-checking fees is not required. If you ignore something due to initially low fee, you lose out to smarter relayers. I don't think it's that difficult to create a smart relayer:

Fair enough. Supporting that on the server does seem the "right" thing to do, and relayers can start with ignoring fee increases at start and only later add that functionality. I am not sure how `hermes` works, but `ts-relayer` checks new packets and handles them right away, it doesn't have some sort of holding pattern for them.

Reply:
i agree with @AdityaSripal here, the user should have the ability to replay the tx with a higher fee.